org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java

```diff
@@ -0,0 +1,295 @@
+package org.eclipse.jdt.internal.core.manipulation;
+/**
+ * This utility class for English words is taken from a blog on CSDN, and we modify it to adapt for the cases of multiply word.
+ * The original blog can be found in https://blog.csdn.net/weixin_44258855/article/details/123201283
```

Comment:
Sorry, this can't be merged into eclipse.org codebase as it refers to a source without clear licensing information, doesnt represent your own code and the snippet itself seem to be largely inadequate to address the problem you are trying to fix.

Reply:
Thanks, I will try to solve this problem soon.

Reply:
@iloveeclipse how did you determine the source of code ? i.e if its devs own or not ? Is there some tool for it ?

Reply:
Are third-party APIs allowed to merge into eclipse.org codebase?

Reply:
> Are third-party APIs allowed to merge into eclipse.org codebase?

See https://www.eclipse.org/projects/handbook/#ip

TL;DR: content added to the jdt.ui bundle must be licensed under EPL license.

Reply:
> @iloveeclipse how did you determine the source of code ? i.e if its devs own or not ? Is there some tool for it ?

I would name it BDA  "brain driven analysis". Of course as every tool that might fail too.
In this concrete context I just assume the "using code posted by someone on some blog page" is most likely description of the code not written by PR author and so it is highly questionable if that is allowed to be merged here.

Reply:
It looks like there is no existing open-source tool written in Java to support the convertion between plural words and singular words.  I will implement my own version soon.

Reply:
@Michaelll123 FWIW: a lot of English words are unnecessary to handle (e.g. geese to goose is unlikely to show up in code and not worth the time to look for it).  There was a pared down version of finding a singular in ConvertLoopOperation.modifyBaseName() which gets a variable name for an enhanced-for-loop (e.g. for (String name : names)).  It handled a number of common usages (e.g. children vs child) and then identified odd English plural suffixes to avoid attempting removing the "s" in which case it defaulted to "element".   You could base off of that and add more direct conversion cases if you want.  Pure plural support is over-kill IMO.

Reply:
@jjohnstn Thanks for the insightful advice! It is indeed a tricky problem to cover all the conversion cases in English, and I will start with ConvertLoopOperation.modifyBaseName().

Reply:
Thanks for all your valuable advice, and I have finished the requested changes. Please kindly review the latest pull request.